pkg/controller/controller_binding.go

```diff
@@ -347,38 +350,69 @@ func isPlanBindable(serviceClass *v1alpha1.ServiceClass, plan *v1alpha1.ServiceP
 
 func (c *controller) injectBinding(binding *v1alpha1.Binding, credentials map[string]interface{}) error {
 	glog.V(5).Infof("Creating Secret %v/%v", binding.Namespace, binding.Spec.SecretName)
-	secret := &v1.Secret{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      binding.Spec.SecretName,
-			Namespace: binding.Namespace,
-		},
-		Data: make(map[string][]byte),
-	}
 
+	secretData := make(map[string][]byte)
 	for k, v := range credentials {
 		var err error
-		secret.Data[k], err = serialize(v)
+		secretData[k], err = serialize(v)
 		if err != nil {
 			return fmt.Errorf("Unable to serialize credential value %q: %v; %s",
 				k, v, err)
 		}
 	}
 
-	found := false
+	secretClient := c.kubeClient.Core().Secrets(binding.Namespace)
+	// Retry loop for upserting the Secret
+	for {
+		found := true
+		existingSecret, err := secretClient.Get(binding.Spec.SecretName, metav1.GetOptions{})
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				return err
+			}
+			found = false
+		}
 
-	_, err := c.kubeClient.Core().Secrets(binding.Namespace).Get(binding.Spec.SecretName, metav1.GetOptions{})
-	if err == nil {
-		found = true
-	}
+		if found {
+			if !IsOwnedBy(existingSecret, binding) {
+				return fmt.Errorf("Secret '%s' is not owned by Binding '%s', ownerReferences: %v",
+					existingSecret.Name, binding.Name, existingSecret.OwnerReferences)
+			}
+			existingSecret.Data = secretData
+			_, err = secretClient.Update(existingSecret)
+			if errors.IsConflict(err) {
+				// Conflicting update detected, try again
+				continue
+			}
+		} else {
+			secret := &v1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      binding.Spec.SecretName,
+					Namespace: binding.Namespace,
+					OwnerReferences: []metav1.OwnerReference{
+						*NewControllerRef(binding, &bindingControllerKind),
+					},
+				},
+				Data: secretData,
+			}
+			_, err = secretClient.Create(secret)
+			if errors.IsAlreadyExists(err) {
+				// Concurrent controller has created secret under the same name,
+				// Update the secret at the next iteration
+				continue
+			}
+		}
 
-	if found {
-		_, err = c.kubeClient.Core().Secrets(binding.Namespace).Update(secret)
-	} else {
-		_, err = c.kubeClient.Core().Secrets(binding.Namespace).Create(secret)
+		// Escape from the loop in case of non-retryable error
+		if err != nil {
+			return err
+		}
+		// Success
+		break
 	}
 
-	if err != nil || binding.Spec.AlphaPodPresetTemplate == nil {
-		return err
+	if binding.Spec.AlphaPodPresetTemplate == nil {
```

Comment:
This and below will be deleted soon (see https://github.com/kubernetes-incubator/service-catalog/issues/994), thus I didn't bother fixing this part